src/g3log/logmessage.hpp

```diff
@@ -87,7 +88,8 @@ namespace g3 {
       // Complete access to the raw data in case the helper functions above
       // are not enough.
       //
-      std::time_t _timestamp;
+      // TODO: unique_ptr should be considered instead of shared_ptr
+      std::shared_ptr<Timestamp> _timestamp;
```

Comment:
_timestamp can be either a pointer (raw / unique / shared) or a template because different TimestampProvider's return _timestamp's of different types.

As you pointed out smart pointer brings a bit of overhead during runtime. Template will have impact for compile time, which I think is less critical. I am not sure whether raw pointers will cause cache misses.

So which approach do you think is more suitable in this case?